### ISSUE TITLE
Error: Conflict on chrome.channel: "stable" !== "dev"

### ISSUE THREAD

**Issue by pimterry** [vweevers/win-detect-browsers#55]
Hi,

I'm seeing an issue in [my app](http://httptoolkit.tech/mock), where a user with the Chrome dev channel installed sees this error:

> Error: Conflict on chrome.channel: "stable" !== "dev"

User report is here: https://github.com/httptoolkit/feedback/issues/38

This is coming from the browser detection in [here](https://github.com/vweevers/win-detect-browsers/blob/022034f5cb19b615946d4b7619701ca6274a5055/lib/finder.js#L218), but I'm not 100% clear why this would happen. Are there any pointers you can give me?

Happy to look into it more myself and send a PR once I've worked it out, but any pointers in the right direction would be very appreciated!

**Comment by vweevers** [comment]
IIRC it happens when a distinct browser (by path) is found via two avenues, and one avenue detected a stable channel, while the other dectected a dev channel.

**Comment by vweevers** [comment]
The output of `npx win-detect-browsers --debug` would help.

**Comment by LachlanArthur** [comment]
Hi, I'm the user that reported the issue.

Here's the output of that debug command:

<details>
<summary>

`Work@Sobek:~ $ npx win-detect-browsers --debug`

</summary>

```
  win-detect-browsers Found chrome:
  win-detect-browsers   - C:\Users\Work\AppData\Local\Google\Chrome SxS\Application\chrome.exe
  win-detect-browsers   @ HKCU\Software\Google\Update /v LastInstallerSuccessLaunchCmdLine +0ms
  win-detect-browsers chrome: win-version-info failed for C:\Users\Work\AppData\Local\Google\Chrome SxS\Application\chrome.exe Error: win-version-info is unable to access file: C:\Users\Work\AppData\Local\Google\Chrome SxS\Application\chrome.exe
    at module.exports (C:\Users\Work\AppData\Roaming\npm-cache\_npx\6840\node_modules\win-detect-browsers\node_modules\win-version-info\index.js:17:22)
    at annotate (C:\Users\Work\AppData\Roaming\npm-cache\_npx\6840\node_modules\win-detect-browsers\lib\version-info.js:20:16)
    at Finder.found (C:\Users\Work\AppData\Roaming\npm-cache\_npx\6840\node_modules\win-detect-browsers\lib\finder.js:225:15)
    at registry.value (C:\Users\Work\AppData\Roaming\npm-cache\_npx\6840\node_modules\win-detect-browsers\lib\finder.js:149:21)
    at process._tickCallback (internal/process/next_tick.js:152:19)
    at Function.Module.runMain (module.js:705:11)
    at findNodeScript.then.existing (C:\Users\Work\scoop\persist\nvm\nodejs\v9.6.1\node_modules\npx\node_modules\libnpx\index.js:268:14)
    at <anonymous> +0ms
  win-detect-browsers Found chrome:
  win-detect-browsers   - C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
  win-detect-browsers   @ HKLM\Software\Wow6432Node\Microsoft\Windows\CurrentVersion\App Paths\chrome.exe +13ms
  win-detect-browsers Found chrome:
  win-detect-browsers   - C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
  win-detect-browsers   @ HKLM\Software\Microsoft\Windows\CurrentVersion\App Paths\chrome.exe +4ms
  win-detect-browsers Found chrome:
  win-detect-browsers   - C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
  win-detect-browsers   @ HKLM\Software\Wow6432Node\Clients\StartMenuInternet\Google Chrome\shell\open\command +4ms
  win-detect-browsers Found chrome:
  win-detect-browsers   - C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
  win-detect-browsers   @ HKLM\Software\Clients\StartMenuInternet\Google Chrome\shell\open\command +5ms
  win-detect-browsers Found chrome (canary) (64):
  win-detect-browsers   - C:\Users\Work\AppData\Local\Google\Chrome SxS\Application\chrome.exe
  win-detect-browsers   @ Software\Google\Update\ClientState\{4EA16AC7-FD5A-47C3-875B-DBF4A2008C20}\LastInstallerSuccessLaunchCmdLine +4ms
  win-detect-browsers Found firefox:
  win-detect-browsers   - C:\Program Files (x86)\Mozilla Firefox\firefox.exe
  win-detect-browsers   @ HKLM\Software\Wow6432Node\Clients\StartMenuInternet\firefox.exe\shell\open\command +6ms
  win-detect-browsers Found firefox:
  win-detect-browsers   - C:\Program Files (x86)\Mozilla Firefox\firefox.exe
  win-detect-browsers   @ HKLM\Software\Clients\StartMenuInternet\firefox.exe\shell\open\command +4ms
  win-detect-browsers Found firefox:
  win-detect-browsers   - C:\Program Files (x86)\Mozilla Firefox\firefox.exe
  win-detect-browsers   @ HKCU\Software\Clients\StartMenuInternet\firefox.exe\shell\open\command +3ms
  win-detect-browsers firefox: found key "69.0 (x64 en-GB)" in HKLM\Software\Mozilla\Mozilla Firefox /v CurrentVersion +3ms
  win-detect-browsers firefox: found key "69.0 (x64 en-GB)" in HKCU\Software\Mozilla\Mozilla Firefox /v CurrentVersion +2ms
  win-detect-browsers Found ie:
  win-detect-browsers   - C:\Program Files\Internet Explorer\iexplore.exe
  win-detect-browsers   @ HKLM\Software\Wow6432Node\Clients\StartMenuInternet\iexplore.exe\shell\open\command +1ms
  win-detect-browsers Found ie:
  win-detect-browsers   - C:\Program Files\Internet Explorer\iexplore.exe
  win-detect-browsers   @ HKLM\Software\Clients\StartMenuInternet\iexplore.exe\shell\open\command +3ms
  win-detect-browsers Found opera:
  win-detect-browsers   - C:\Program Files (x86)\Opera\Launcher.exe
  win-detect-browsers   @ HKLM\Software\Wow6432Node\Clients\StartMenuInternet\OperaStable\shell\open\command +3ms
  win-detect-browsers Found opera:
  win-detect-browsers   - C:\Program Files (x86)\Opera\Launcher.exe
  win-detect-browsers   @ HKLM\Software\Clients\StartMenuInternet\OperaStable\shell\open\command +5ms
  win-detect-browsers Found opera:
  win-detect-browsers   - C:\Program Files (x86)\Opera\Launcher.exe
  win-detect-browsers   @ HKLM\Software\Classes\OperaStable\shell\open\command +3ms
  win-detect-browsers Found opera:
  win-detect-browsers   - C:\Program Files (x86)\Opera\Launcher.exe
  win-detect-browsers   @ HKCU\Software\Classes\OperaStable\shell\open\command +3ms
  win-detect-browsers Found chrome (stable):
  win-detect-browsers   - C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
  win-detect-browsers   @ Software\Classes\ChromeHTML\shell\open\command +3ms
  win-detect-browsers Found firefox:
  win-detect-browsers   - C:\Program Files (x86)\Mozilla Firefox\firefox.exe
  win-detect-browsers   @ HKLM\Software\Mozilla\Mozilla Firefox\69.0 (x64 en-GB)\Main /v PathToExe +3ms
  win-detect-browsers Found chrome:
  win-detect-browsers   - C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
  win-detect-browsers   @ default location +4ms
  win-detect-browsers Found firefox:
  win-detect-browsers   - C:\Program Files (x86)\Mozilla Firefox\firefox.exe
  win-detect-browsers   @ default location +3ms
  win-detect-browsers Found firefox:
  win-detect-browsers   - C:\Program Files\Firefox Nightly\firefox.exe
  win-detect-browsers   @ default location +3ms
  win-detect-browsers Found ie:
  win-detect-browsers   - C:\Program Files (x86)\Internet Explorer\iexplore.exe
  win-detect-browsers   @ default location +4ms
  win-detect-browsers Found ie:
  win-detect-browsers   - C:\Program Files\Internet Explorer\iexplore.exe
  win-detect-browsers   @ default location +5ms
  win-detect-browsers Found opera:
  win-detect-browsers   - C:\Program Files (x86)\Opera\Launcher.exe
  win-detect-browsers   @ default location +5ms
  win-detect-browsers Found chrome (dev) (64):
  win-detect-browsers   - C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
  win-detect-browsers   @ Software\Wow6432Node\Google\Update\ClientState\{8A69D345-D564-463C-AFF1-A69D9E530F96}\UninstallString +3ms
C:\Users\Work\AppData\Roaming\npm-cache\_npx\6840\node_modules\win-detect-browsers\lib\finder.js:218
            throw new Error(`Conflict on ${this.name}.${k}: ${v1} !== ${v2}`)
            ^

Error: Conflict on chrome.channel: "stable" !== "dev"
    at Finder.found (C:\Users\Work\AppData\Roaming\npm-cache\_npx\6840\node_modules\win-detect-browsers\lib\finder.js:218:19)
    at fs.access (C:\Users\Work\AppData\Roaming\npm-cache\_npx\6840\node_modules\win-detect-browsers\lib\chrome\find-update-clients.js:75:25)
    at FSReqWrap.oncomplete (fs.js:153:20)
```

</details>

I used to have Canary as well, but it has been uninstalled.

**Comment by vweevers** [comment]
I think the problem is that one of the detection methods finds Chrome dev in the registry at `Software\Classes\ChromeHTML\shell\open\command` and wrongly assumes the `ChromeHTML` part (rather than `ChromeSSHTM.*`, which is canary) means stable channel:

https://github.com/vweevers/win-detect-browsers/blob/d45d9383444ed79db254b89eb6a04edbd7adeebf/lib/chrome/find-progids.js#L32

Perhaps that should be `id === 'ChromeHTML' ? null : 'canary'`.

**Comment by pimterry** [comment]
@vweevers I've been looking into this, as I've hit another similar issue (https://github.com/httptoolkit/feedback/issues/46).

I don't quite have a proper fix, but I've been looking at short-term mitigations. Is there a reason that [this error](https://github.com/vweevers/win-detect-browsers/blob/master/lib/finder.js#L218) is thrown directly, rather than passing it to the callback? Right now this error crashes my process, but if it was possible to handle it then I could avoid that.

**Comment by vweevers** [comment]
> Is there a reason that this error is thrown directly

Yes, because I consider it a bug in `win-detect-browsers`.

I think the proper fix is this (https://github.com/vweevers/win-detect-browsers/issues/55#issuecomment-535793110):

> Perhaps that should be id === 'ChromeHTML' ? null : 'canary'.

I don't have the bandwidth atm to think about implications. But if you could send a PR, then i'll make a prerelease, and look into it more closely at a later time.

**Comment by vweevers** [comment]
Fixed by #61.
